sound/soc/sof/ipc4-pcm.c

```diff
@@ -216,6 +296,27 @@ static int sof_ipc4_trigger_pipelines(struct snd_soc_component *component,
 	if (!pipeline_list->pipelines)
 		return 0;
 
+	/*
+	 * Every pipeline of a chained DMA stream should have use_chain_dma attribute, so
+	 * this loop is only necessary for checking the correctness of the topology. The loop
+	 * together with the check in sof_ipc4_chain_dma_trigger() completes the check.
+	 */
+	for (i = pipeline_list->count - 1; i >= 0; i--) {
+		struct snd_sof_widget *pipe_widget;
+		struct sof_ipc4_pipeline *pipeline;
+
+		spipe = pipeline_list->pipelines[i];
+		pipe_widget = spipe->pipe_widget;
+		pipeline = pipe_widget->private;
+
+		/*
+		 * If any pipeline in the list has use_chain_dma attribute we proceed to
+		 * chained DMA trigger function that handles the rest for the substream.
+		 */
+		if (pipeline->use_chain_dma)
+			return sof_ipc4_chain_dma_trigger(sdev, pipeline_list, state, cmd);
+	}
```

Comment:
I guess there is no way to reduce the need for the loop by checking first the `[0]->pipe_widget->use_chain_dma` as it is plausible that some of the pipelines in the chain might have the flag set, making the path unworkable?

Reply:
Like the comments indicate, the check you suggest would be enough for the functionality. However, this loop is there to verify the use_chain_dma attributes are used consistently in the topology. e.g. handle every pipeline_list with even a single use_chain_dma attribute with sof_ipc4_chain_dma_trigger(), which in turn makes sure all the pipelines in the list have the attribute, and produces an appropriate error print if they don't.